### PR TITLE
Site templates feature

### PR DESCRIPTION
**Pull request by jeremy-rifkin**
Currently looks like this:

image
image

Still need to:
- [x] Figure out a better way to size the image preview so it doesn't overflow
- [x] Better deploy static images
- [x] Make variations of screenshots for the various themes?
- [ ] More templates

Future features can include favoriting templates and user-defined templates

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -764,6 +767,7 @@ async function main() {
             staticHeaders(res);
             res.end('User-agent: *\nSitemap: https://godbolt.org/sitemap.xml\nDisallow:');
         })
+        .use('/template_screenshots', express.static('etc/template_screenshots/'))
```
Comment (jeremy-rifkin) [inline app.js:770]: Is there a better way to serve static assets?

Comment (partouf) [reply app.js:770]: This is good for local users, I think. We'd probably upload the templates to our CDN on S3, so there probably needs to be a setting that swaps the URL to `https://static.ce-cdn.net/template_screenshots/...`

Comment (jeremy-rifkin) [reply app.js:770]: What's needed to deploy the images to the cdn?

Comment (partouf) [reply app.js:770]: It's complicated. Our static files are zipped up and uploaded to S3 in our github workflow. (here https://github.com/compiler-explorer/compiler-explorer/blob/main/.github/workflows/test-and-deploy.yml#L76)

But then only once we decide to deploy the version, the static files will be uploaded to the CDN using this function: https://github.com/compiler-explorer/infra/blob/33265d8edd518a31d96cbf36ee3f4345204f24ef/bin/lib/cli/builds.py#L49

Comment (jeremy-rifkin) [reply app.js]: Thanks, I've updated how images are handled. Now using the same webpack magic used for the language icons 🤞

Comment (partouf) [reply app.js]: Didn't realize that was possible, well done

**Comment by jeremy-rifkin** [comment]
> See you've added only site-templates.properties, would it make sense to make .defaults, then link symlink .amazon to .defaults so that local users can override with .local?

I updated the config file to not be a .properties file as it doesn't go through the existing properties system. Might be worth supporting a .local override but I think I'd defer that to later iterations

File: static/widgets/site-templates-widget.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    async getTemplates() {
+        if (this.templatesConfig === null) {
+            this.templatesConfig = await new Promise((resolve, reject) => {
+                $.getJSON(window.location.origin + window.httpRoot + 'api/getSiteTemplates', resolve);
+            });
+        }
+        return this.templatesConfig as siteTemplatesType;
+    }
+    getCurrentTheme() {
+        return Settings.getStoredSettings()['theme'];
```
Comment (jeremy-rifkin) [inline static/widgets/site-templates-widget.ts:48]: Is there a better way to do this?

**Comment by partouf** [comment]
I wonder if it would be better if you got rid the header in the screenshots. I think it should be possible if you use `https://godbolt.org/e/`?

**Comment by jeremy-rifkin** [comment]
Good idea :)

File: lib/handlers/api.js [hunk 3]
```diff
@@ -117,6 +118,12 @@ export class ApiHandler {
                 res.send(all);
             })
             .all(methodNotAllowed);
+        this.handle
+            .route('/getSiteTemplates')
```
Comment (partouf) [inline lib/handlers/api.js:122]: probably better to call this `/siteTemplates`

**Comment by jeremy-rifkin** [comment]
Should be ready to merge whenever, happy to give it more time for any other comments/reviews

**Review by partouf — APPROVED** [review]

**Review by matsjla — APPROVED** [review]

**Comment by partouf** [comment]
this is now live

**Pull request merged**